### ISSUE TITLE
javac is not respecting the JAVA_HOME and javaHome

### ISSUE THREAD

**Issue by fommil** [sbt/sbt#1519]
related to
- https://github.com/sbt/sbt/issues/510
- https://github.com/paulp/sbt-extras/issues/83

When sbt is started with a specific version of JAVA_HOME, the `javac` is still picked up by PATH rather than `JAVA_HOME/bin/javac`. This is both unexpected and can lead to bad binaries being built (e.g. java 7 classfiles produced in a java 6 project).

Global solutions as proposed in #510 are non useful because the need to change the java version is the exception case, not the general rule (and this sort of information cannot appear in project SCM).

**Comment by fommil** [comment]
btw, I work around this with these lines at the top of my sbt script https://github.com/fommil/unix/blob/master/bin/sbt

```
if [ -e ".javahome" ]; then
  export JAVA_HOME=`cat .javahome`
  export JDK_HOME=$JAVA_HOME
  export PATH=$JAVA_HOME/bin:$PATH
fi
```

**Comment by pdolega** [comment]
@fommil We had the same problem. After quick debugging session I got an impression that it has been fixed in 0.13.8.

**Comment by fommil** [comment]
Cool! This has bothered me for years.

**Comment by fommil** [comment]
Are you sure it's using the correct javac?

**Comment by pdolega** [comment]
It appears so. I got codebase that uses Java 8 lambdas and following seems to work:

```
[pdolega@pdolega istudent]$ javac -version
javac 1.7.0_51
[pdolega@pdolega istudent]$ java -version
java version "1.7.0_51"
Java(TM) SE Runtime Environment (build 1.7.0_51-b13)
Java HotSpot(TM) 64-Bit Server VM (build 24.51-b03, mixed mode)
[pdolega@pdolega istudent]$ echo $JAVA_HOME
/usr/java/jdk1.8.0_11
[pdolega@pdolega istudent]$ ./activator clean compile
...
...
[success] Total time: 70 s, completed May 22, 2015 9:53:01 AM
```

Running it again after doing:

```
[pdolega@pdolega istudent]$ JAVA_HOME=$JAVA_HOME7
[pdolega@pdolega istudent]$ echo $JAVA_HOME
/usr/java/jdk1.7.0_51
```

fails on compilation (which is what should happen as codebase uses Java 8 source).

**Comment by fommil** [comment]
I don't know what your `activator` script does, but if it's calling out to `sbt` then this is looking promising.

**Comment by fommil** [comment]
although, `JDK_HOME` (JDK) should be respected before `JAVA_HOME` (JRE or JDK).

**Comment by pdolega** [comment]
@fommil activator basically runs build.sbt. Also I did not check JDK_HOME - our goal was to run multiple different builds from single Jenkins instance and, as some of builds require different Java version than the other, actual usage of JAVA_HOME (instead of javac on path) was enough to make everything work.

**Comment by fommil** [comment]
sweet

**Comment by fommil** [comment]
now all we need is a way to set the java-home on a per-project basis (like paulp's script allows)

**Comment by pdolega** [comment]
On the side note it's hard to believe this was marked as `enhacement` (it sure seems like bug to me). 

@jsuereth it seems like the core part of this is issue is fixed - it may be worth to close it. Just saying.

Anyway, thanks. Over 'n out.

**Comment by fommil** [comment]
this is not fixed in 0.13.9. I can confirm that the wrong version of javac is if `javaHome` is set correctly.

**Comment by dwijnand** [comment]
Fixed in https://github.com/sbt/sbt-launcher-package/pull/112
